### DISCUSSION TITLE
Cache not cleared correctly when using {% cache until %} (CraftCMS 3.5.17.1)

### DISCUSSION THREAD

**Discussion by jannisborgers** [craftcms/cms#13134] [Q&A] [answered]
We have a theater site with a large event list. The events are rows in a table field inside entries from a “production“ section (190 entries in that section, 230 events across these entries).

On the homepage, we display events for the next 3 days that have events. The program page displays all events.

In both cases, only events today or in the future are displayed. Past events should be hidden.

**Overview**
1. The first request on any given day queries all events today or in the future (this seems to work correctly in my tests)
2. The result is cached and will be shown all day (this works, Yii shows ~40 DB queries instead of the original ~400)
3. At 24:00:00, the cache expires
4. Back to step 1 for the first request after cache expiry

**Problem**
Steps 3 and 4 don’t seem to work consistently: This morning (April 29), the next event shown on the homepage was for April 29 (correct). At the same time, the program page still showed the event for yesterday, April 28. They use the same template, only slicing the next event days in case of the homepage.

Any ideas if my caching plan is flawed? Is there a chance that old results be cached? Or could the cache not expire as planned?

**_layout.twig**
```twig
{% set startOfToday = now|date_modify("00:00:01") %}
{% set endOfToday = now|date_modify("23:59:59") %}
```

**eventList.twig:**
```twig
{# Cache event list until 23:59:59 today #}
{% cache until endOfToday %}

    {# stillRunning is a preparse field used to pre-filter only productions that had events in the future when they were last saved are queried. When a production entry is saved and all its events are in the past, stillRunning will be 0. This reduces the initial query because there are only around 30 active productions at any given time of the year. #}
    {# productionEvents is a table field that has date and time field (this version of Craft didn’t have dateTime fields yet) #}
    {% set allProductions = craft.entries()
        .section("productions")
        .stillRunning("1")
        .with(["productionEvents:date", "productionEvents:time"])
        .all() 
    %}

    {# Merge date and time fields per event per production #}
   {# Add all events to list that are after 00:00:01 today #}
    {% set allEvents = [] %}
    {% for production in allProductions %}
        {% for event in production.productionEvents %}
            {% set eventDate = event.date|date_modify(event.time|time()) %}
            {% if eventDate > startOfToday %}
                {% set allEvents = allEvents|merge([event]) %}
            {% endif %}
        {% endfor %}
        {% set allEvents = allEvents|sort((a, b) => a.date|date_modify(a.time|time()) > b.date|date_modify(b.time|time())) %}
    {% endfor %}

    {# Output list here #}
    {% for event in allEvents %}
        {# … #}
    {% endfor %}

{% endcache %}
```

**Comment by vnali** [comment, marked as answer]
@jannisborgers I remember I had same issue with cache until on craft 3 and it resolved later on 3.6.12.1 https://github.com/craftcms/cms/issues/7826

**Reply by jannisborgers** [reply]
@vnali Thank you very much for the tip! I upgraded the system to the latest version and will keep an eye on it to make sure it’s not only working locally. I’ll update this post if the issue is solved.

**Reply by brandonkelly** [reply]
@jannisborgers to be clear, were you on an earlier version than 3.6.12.1?

**Reply by jannisborgers** [reply]
@brandonkelly Yes that is correct, this site was still on 3.5.17.1 before the original issue was found and fixed (via the issue linked above).

**Reply by jannisborgers** [reply]
To clarify: I upgraded the site to 4.4.8 today, not to 3.6.12.1.

**Reply by brandonkelly** [reply]
Thanks! Will close this, assuming it’s the same issue as #7826. Let me know if it crops up again though!

**Reply by jannisborgers** [reply]
I can confirm that the cache clears correctly now using the template I posted above.
